### DISCUSSION TITLE
[Feature Request] Ability to add rulesets and filterlists as sources.

### DISCUSSION THREAD

**Discussion by LucioAmely** [DNSCrypt/dnscrypt-proxy#2530] [Ideas]
The sources and list fetching feature is already implemented in dnscrypt-proxy. Please provide us a convenient way to automatically fetch rulesets and filterlists on all supported platforms.

1.) There will be 2 types of ruleset and filterlist files. One appended with `-local` and another appended with `-remote`. Local files will have priority over remote files.

2.) Users will configure remote rulesets and filterlist sources from `dnscrypt-proxy.toml` together with their own refresh delays.

3.) At app start dnscrypt-proxy will initialize connection to the DNS server using previous rulesets and filterlists if needed. (I do not know if they are being used or needed during the initial connection phase.)

4.) Then dnscrypt-proxy will initiate downloading of rulesets and filterlists configured in the `dnscrypt-proxy.toml` file and will queue (if queuing not possible then will block) all requests sent to the app until the lists are downloaded and initialized.

**Comment by LucioAmely** [comment]
@jedisct1 May I ask why #2535 got deleted without any conversation?

**Reply by jedisct1** [reply]
Because there was no content. Just a link to this page that only contained your wishlist, without any consensus nor discussion.

**Reply by LucioAmely** [reply]
Then please let's discuss it's ups and downs here. I have waited for 4 days for any reply to my thread but there was none. I am not sure how long should I wait for an open discussion.

**Reply by jedisct1** [reply]
"Please provide us" is not a decent way to ask.

**Reply by LucioAmely** [reply]
Sorry for that. Didn't want to be intentionally mean.

**Comment by jedisct1** [comment]
Setting up a webserver in order to use it to fetch some configuration files sounds a bit cumbersome, and I'm not sure a lot of people would actually use that. People with such a need and skills can use tools such as `curl`, `rsync` or `mutagen` to sync  configuration files.

Right now, a restart is required after a change in a configuration file. This is mostly transparent to clients since they will all automatically retry after a timeout. However, after a restart, the cache is lost and the server selection has to be made again, so for a couple seconds, performance may not be optimal.

Watching files for changes and reparsing them after every change would solve this. There was an early experiment to do so, but it didn't work reliably.

These files must not be loaded while they are being edited. In order to achieve that, Unix tools often expect a `HUP` signal to be received in order to reload configuration files. Unfortunately, that doesn't work under Windows.

For blocked and allowed names, the `generate-domains-blocklist.py` script can add a checksum at the end of the file, so we know if the file is consistent or not.

For blocked and allowed IP and IPv6 addresses, maybe a distinct script can be written. Something that merges overlapping IP addresses, but also expands CIDRs.

On invalid configurations, such as blocked names with references to unknown time ranges, I guess we should keep the previous configuration and emit errors.

**Comment by lifenjoiner** [comment]
The key is reloading.
So, my solution is a bit different, not adapting those rules as remote sources, but just reloading local rules. And the soul is delegating all the transformations to a [script file](https://github.com/lifenjoiner/dnsforwarder/blob/25fbaef2fb4701fba64d32e786b254e300d7774c/default.en.config#L311).

In details:
1. [Define which rules would be reloaded, and the interval](https://github.com/lifenjoiner/dnsforwarder/blob/25fbaef2fb4701fba64d32e786b254e300d7774c/default.en.config#L351-L376).
2. Write the [script](https://github.com/lifenjoiner/dnsforwarder/blob/25fbaef2fb4701fba64d32e786b254e300d7774c/default.en.config#L311). Make sure all the rules will be updated successfully, or fall back to backups.
3. At starting up, just load the existing rule files.
4. Run the script to update the rules ASAP.
5. When the script finished, reload the rules that defined in step 1.
6. Sleep for the interval, and then repeat from step 4.

The external script can be vary file formats (bat, ps1, sh, js, py and etc.), and it can do vary transformations. It is more flexible than doing the work by the main executable itself.
